src/libsyntax/parse/parser.rs

```diff
@@ -5401,23 +5401,26 @@ impl<'a> Parser<'a> {
             }
             let mut err = self.diagnostic().struct_span_err(id_sp,
                 "cannot declare a new module at this location");
-            let this_module = match self.directory.path.file_name() {
-                Some(file_name) => file_name.to_str().unwrap().to_owned(),
-                None => self.root_module_name.as_ref().unwrap().clone(),
-            };
-            err.span_note(id_sp,
-                          &format!("maybe move this module `{0}` to its own directory \
-                                     via `{0}/mod.rs`",
-                                    this_module));
+            if id_sp != syntax_pos::DUMMY_SP {
+                let full_path = self.sess.codemap().span_to_filename(id_sp);
+                let path = Path::new(&full_path);
+                let filename = path.file_stem().unwrap();
+                let parent = path.parent().unwrap_or(Path::new(""))
+                                          .to_str().unwrap_or("").to_owned();
+                let path = format!("{}/{}",
+                                   if parent.len() == 0 { "." } else { &parent },
```

Comment:
out of curiosity, can you construct an example where the `parent.len()` is in fact equal to zero? I would have thought the context where this is arising will ensure that there is always a parent, but maybe I am missing a scenario.